### PR TITLE
Register dockerd.exe as message file for iotedge-moby

### LINKED ISSUE #674: Events from iotedge-moby event source on Windows cannot be properly queried due to missing EventMessageFile
## Expected Behavior
The following command should return any relevant events logged by the iotedge-moby service instance of dockerd.exe:

Get-WinEvent -ea SilentlyContinue -FilterHashtable @{ProviderName='iotedge-moby';LogName='application';StartTime=[datetime]::Now.AddMinutes(-5)}

## Current Behavior
The Get-WinEvent command above returns the following error:

> Get-WinEvent : The parameter is incorrect
> At line:1 char:5
> \+     Get-WinEvent -ea SilentlyContinue -FilterHashtable @{ProviderName ...
> \+     \~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     \+ CategoryInfo          : NotSpecified: (:) [Get-WinEvent], EventLogException
>     \+ FullyQualifiedErrorId : System.Diagnostics.Eventing.Reader.EventLogException,Microsoft.PowerShell.Commands.GetWinEventCommand

In addition, looking in the Application event log with event viewer shows something like this:

> The description for Event ID 1 from source iotedge-moby cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.
> 
> If the event originated on another computer, the display information had to be saved with the event.
> 
> The following information was included with the event: 
> 
> 824569b760f7f05239a531bc9a98212447feefac4a68a0ee501cdb6828e69950 cleanup: failed to delete container from containerd: no such container

## Steps to Reproduce
1. Install the Azure IoT Edge runtime in a VM running Server 2019 (17763.195) Datacenter Edition according to the instructions at https://docs.microsoft.com/en-us/azure/iot-edge/how-to-install-iot-edge-windows.
2. Once everything is up and running, restart the iotedge-moby service.  At this point, due to another issue, warning and error logs will start being logged by iotedge-moby and will not be queryable via Get-WinEvent.

### Device (Host) Operating System
Windows Server 2019

### Architecture
amd64

### Container Operating System
Windows containers

#### iotedged
iotedge 1.0.5 (d76e0316c6f324345d77c48a83ce836d09392699)

#### Edge Agent
1.0 (logs say it's 1.0.5)

#### Edge Hub
1.0 (logs say it's 1.0.5)

#### Docker
Client:
 Version:           3.0.2
 API version:       1.39 (downgraded from 1.40)
 Go version:        go1.11.1
 Git commit:        e042b58f
 Built:             Tue Dec  4 20:13:19 2018
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       4d60db4
  Built:            Wed Nov  7 00:55:00 2018
  OS/Arch:          linux/amd64
  Experimental:     false

### PR DESCRIPTION
**Pull request by DaveEM**
This fixes #674.  I used dockerd.exe as the message file since that's what Docker for Windows uses when you install it normally.

### CODE REVIEW

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 [hunk 1]
```diff
@@ -857,14 +862,18 @@ function Remove-FirewallExceptions {
...
-    New-ItemProperty "HKLM:\SYSTEM\CurrentControlSet\Services\EventLog\Application\$EdgeEventLogName" `
-        -Name 'TypesSupported' -Value 7 -PropertyType 'DWord' -Force | Out-Null
+function Add-IotEdgeRegistryValues {
+    $regKey = Join-Path $EventLogApplicationRegPath $EdgeEventLogName
+    New-Item $regKey -Force | Out-Null
+    New-ItemProperty $regKey -Name 'CustomSource' -Value 1 -PropertyType 'DWord' -Force | Out-Null
+    New-ItemProperty $regKey -Name 'EventMessageFile' -Value "$EdgeEventLogInstallDirectory\iotedged_eventlog_messages.dll" -PropertyType 'String' -Force | Out-Null
+    New-ItemProperty $regKey -Name 'TypesSupported' -Value 7 -PropertyType 'DWord' -Force | Out-Null
+
+    $regKey = Join-Path $EventLogApplicationRegPath $MobyServiceName
```
Comment (arsing) [inline scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: Only if `$ContainerOs -eq 'Windows'`

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 [hunk 2]
```diff
@@ -857,14 +862,18 @@ function Remove-FirewallExceptions {
...
+    New-Item $regKey -Force | Out-Null
+    New-ItemProperty $regKey -Name 'CustomSource' -Value 1 -PropertyType 'DWord' -Force | Out-Null
+    New-ItemProperty $regKey -Name 'EventMessageFile' -Value "$MobyInstallDirectory\dockerd.exe" -PropertyType 'String' -Force | Out-Null
+    New-ItemProperty $regKey -Name 'TypesSupported' -Value 7 -PropertyType 'DWord' -Force | Out-Null
 
     Write-HostGreen 'Added IoT Edge registry key.'
```
Comment (arsing) [inline scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: `'Added IoT Edge registry keys.'`

**Comment by arsing** [comment]
This will end up causing the same problem that prompted https://github.com/Azure/iotedge/commit/a135bdfb35fc3a163ebeaf7cb211052fa0410a16, namely having an open session for Windows events will prevent upgrade scenarios since `Uninstall-SD` will be unable to delete `dockerd.exe`.

@veyalla @myagley Thoughts?

**Comment by arsing** [comment]
I feel like requiring the user to restart before they can reinstall will be worse, but I'll let the others decide.

**Comment by myagley** [comment]
I talked to @chipalost about this and we decided to merge this PR and accept that uninstall may fail if there is an open event log session. The thinking is that looking at the docker engine logs themselves is generally a rare event when debugging issues on IoT Edge, and having an open event session when attempting to uninstall is even rarer.

**Comment by arsing** [comment]
As we discussed today, I don't know if uninstalling with Event Viewer open is rare.

If we do decide to take this change, at the very least it should also have a message when it fails to delete `dockerd.exe`, similar to when it fails to delete `iotedged_eventlog_messages.dll`. So it should try to delete `"$MobyInstallDirectory\dockerd.exe"` first, and do

```powershell
    elseif ($cmdErr.FullyQualifiedErrorId -ne 'PathNotFound,Microsoft.PowerShell.Commands.RemoveItemCommand') {
        Write-Verbose "$cmdErr"
        Write-HostRed "Could not delete '$MobyInstallDirectory\dockerd.exe'."
        Write-HostRed ('Please close Event Viewer, or any PowerShell windows where you ran Get-WinEvent, ' +
            'then run `Uninstall-SecurityDaemon` again with `-Force`.')
        $success = $false
    }
```

if it couldn't.

**Comment by arsing** [comment]
@veyalla is back in office, so assigning to him for his opinion.

**Pull request merged**